code/game/machinery/doors/door.dm

```diff
@@ -223,62 +260,76 @@
 			else
 				flick("doorc1", src)
 		if("deny")
-			flick("door_deny", src)
-	return
+			if(!stat)
+				flick("door_deny", src)
 
 /obj/machinery/door/proc/open()
 	if(!density)
-		return 1
+		return TRUE
 	if(operating)
 		return
-	if(!ticker)
-		return 0
-	operating = 1
-
+	operating = TRUE
 	do_animate("opening")
 	set_opacity(0)
 	sleep(5)
-	density = 0
+	density = FALSE
 	sleep(5)
-	layer = open_layer
+	layer = initial(layer)
 	update_icon()
 	set_opacity(0)
-	operating = 0
+	operating = FALSE
 	air_update_turf(1)
 	update_freelook_sight()
 
 	// The `addtimer` system has the advantage of being cancelable
 	if(autoclose)
 		autoclose_timer = addtimer(src, "autoclose", normalspeed ? auto_close_time : auto_close_time_dangerous, unique = 1)
 
-	return 1
+	return TRUE
 
 /obj/machinery/door/proc/close()
 	if(density)
-		return 1
-	if(operating)
+		return TRUE
+	if(operating || welded)
 		return
-	operating = 1
+	if(safe)
+		for(var/atom/movable/M in get_turf(src))
+			if(M.density && M != src) //something is blocking the door
+				if(autoclose)
+					addtimer(src, "autoclose", 60)
+				return
+
+	operating = TRUE
 
 	if(autoclose_timer)
 		deltimer(autoclose_timer)
 		autoclose_timer = 0
 
 	do_animate("closing")
-	layer = closed_layer
+	layer = closingLayer
 	sleep(5)
-	density = 1
+	density = TRUE
 	sleep(5)
 	update_icon()
 	if(visible && !glass)
-		set_opacity(1)	//caaaaarn!
+		set_opacity(1)
 	operating = 0
 	air_update_turf(1)
 	update_freelook_sight()
-	return
+	if(safe)
+		CheckForMobs()
+	else
+		crush()
+	return TRUE
+
+/obj/machinery/door/proc/CheckForMobs()
+	if(locate(/mob/living) in get_turf(src))
+		sleep(1)
+		open()
 
 /obj/machinery/door/proc/crush()
 	for(var/mob/living/L in get_turf(src))
+		L.visible_message("<span class='warning'>[src] closes on [L], crushing them!</span>", "<span class='userdanger'>[src] closes on you and crushes you!</span>")

```

Comment:
I see this possibly causing spam issues with doorcrush happy AIs

Reply:
I don't think it'd be that bad as we have the repeated messages combiner in chat, doorcrushing itself is pretty rare